SwiftyJSONAPI/JSONAPIResource.swift

```diff
@@ -102,13 +105,40 @@ public class JSONAPIResource: JSONPrinter {
                 
                 resource.attributes    = includedResource.attributes
                 resource.relationships = includedResource.relationships
+            
                 
-                if !resource.relationships.isEmpty {
+                if !resource.relationships.isEmpty, resource.shouldLoadResource(withCachedResources: &cachedResources) {
+                    
                     resource.parent = self.parent
-                    resource.loadResources(withIncludedResources: includedResources)
+                    resource.loadResources(withIncludedResources: includedResources, cachedResources: &cachedResources)
                 }
+                
                 resource.loaded = .Loaded
             }
         }
     }
 }
+
+private extension JSONAPIResource {
+    
+    // Checks if a resource has been loaded already, prevents bidirectional relationships from being recursively called
+    func shouldLoadResource(withCachedResources cachedResources: inout CachedResources) -> Bool {
+        
+        if var cachedResourceIds = cachedResources[type] {
+            
+            if cachedResourceIds.contains(id) {
+                
+                return false
+            } else {
+                
+                cachedResourceIds.insert(id)
+                cachedResources[type] = cachedResourceIds
```

Comment:
is it a bit opaque/misleading to be updating cachedResources from within this shouldLoad method? sort of an invisible side effect. should we instead add ids to cachedResources once they've actually been loaded?

Reply:
This was the intention to hide away this implementation so the developer will not need to worry about adding the ids themselves.

If we separate it out then we need to repeat the logic of knowing to either add or create a new cached resource.  We can change the sig to make it more clear?

Reply:
sure, and it would still be handled internally, devs wouldn't have to do anything. just suggesting we move the setting/updating of cachedResources out of a helper that tells us whether or not to load a resource, and up to where we are in fact loading that resource...i.e. into the
```if !resource.relationships.isEmpty, resource.shouldLoadResource...```
conditional above